reference/src/representation/structs-and-tuples.md

```diff
@@ -0,0 +1,349 @@
+fields are declared in the struct.
+
+However, that rule is not true today. For example, the compiler
+includes an option (called "optimization fuel") that will enable us to
+alter the layout of only the "first N" structs declared in the
+source. When one is accidentally relying on the layout of a structure,
+this can be used to track down the struct that is causing the problem.
+
+[#42877]: https://github.com/rust-lang/rust/issues/42877
+[pg-unsized-tuple]: https://play.rust-lang.org/?gist=46399bb68ac685f23beffefc014203ce&version=nightly&mode=debug&edition=2015
+
+There are also benefits to having fewer guarantees. For example:
+
+- Code hardening tools can be used to randomize the layout of individual structs.
+- Profile-guided optimization might analyze how instances of a
+particular struct are used and tweak the layout (e.g., to insert
+padding and reduce false sharing).
+    - However, there aren't many tools that do this sort of thing
+([1](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-420650851),
+[2](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-420681763)). Moreover,
+it would probably be better for the tools to merely recommend
+annotations that could be added
+([1](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-420077105),
+[2](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-420077105)),
+such that the knowledge of the improved layouts can be recorded in the
+source.
+
+As a more declarative alternative, @alercah [proposed a possible
+extension](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/12#issuecomment-420165155)
+that would permit one to declare that the layout of two structs or
+types are compatible (e.g., `#[repr(as(Foo))] struct Bar { .. }`),
+thus permitting safe transmutes (and also ABI compatibility).  One
+might also use some weaker form of `#[repr(C)]` to specify a "more
+deterministic" layout. These areas need future exploration.
+
+## Counteropinions and other notes
+
+@joshtripplet [argued against reordering struct
+fields](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-417953576),
+suggesting instead it would be better if users reordering fields
+themselves. However, there are a number of downsides to such a
+proposal (and -- further -- it does not match our existing behavior):
+
+- In a generic struct, the [best ordering of fields may not be known
+  ahead of
+  time](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-420659840),
+  so the user cannot do it manually.
+- If layout is defined, then it becomes part of your API, such taht
+  reordering fields is a breaking change for your clients (if we
+  consider unsafe code that may rely on the layout, then this applies
+  [even to structs with named
+  fields](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-420117856).
```

Comment:
Implementation details like struct layout are not part of your API and don't have to be semver-maintained *unless you say so*. unsafe code can observe all sorts of things, as there is no way to guarantee compatibility with that.

So, I disagree with the implication stated here.

Reply:
Fields are listed in declaration order in rustdoc, so *if all fields are public* and the struct isn't `non_exhaustive` it seems rather reasonable for a client to assume that their relative order is intended to be exposed (just as the number and names and types of public fields is already implicitly part of the API). One can certainly argue that they shouldn't but I don't think it's realistic to establish such a convention at this point, and since the threat here is UB rather than just code no longer compiling after a dependency bump, it seems risky to try.

Reply:
> One can certainly argue that they shouldn't but I don't think it's realistic to establish such a convention at this point,

Are you talking here about `repr(Rust)` structs as well? As in, the compiler is able to reorder their fields, except if all fields are `pub` ? That would be a bit of subtle knowledge that one would need to know about the usage of `pub` on struct fields. I think I would prefer a more explicit way of telling the consumers of a struct that they can rely on the relative order of the fields, like `repr(C)`.

Reply:
This is in the context of the part of the text we're commenting on here, i.e., we are talking about a scenario where we guarantee the layout of all repr(Rust) structs. The question is then, when will consumers of another library's type be allowed to rely on the declaration order of the fields: implicitly whenever the fields are public, or only if it is explicitly guaranteed in the documentation? Certainly if fields are private, there's no reasonable expectation about their presence let alone relative order, but for reasons outlined above I think it would be user-hostile to say "the docs tell you which order these public fields are declared in but you need to forget about this and not use that knowledge".

Reply:
I will rewrite to clarify.